Source Cluster Providers from New "cluster-up" Repository

davidvossel opened the pull request:
# Overview (and philosophy behind this change)

Our development and CI flow is split into two separate steps when setting up a cluster. There's the ```make cluster-up``` command which brings up the cluster, and there's the ```make cluster-sync``` command which installs a build of KubeVirt based on the code in the source tree.

Multiple projects have been replicating the kubevirt logic around ```make cluster-up``` while applying their own custom logic to ```make cluster-sync```. A pattern has emerged where it's obvious that ```make cluster-up``` should be something generic the entire KubeVirt ecosystem can reuse, while each project should be responsible for their own ```make cluster-sync``` logic to install their src tree builds into the cluster.

## High Level Changes
* kubevirt/kubevirt is now only responsible for ```make cluster-sync``` logic
* The ```cluster/``` directory is now sourced from https://github.com/kubevirt/kubevirtci using a specific git hash.
* Everything not related to ```make cluster-up``` in the previous cluster/ directory has been split into separate locations in the ```hack/``` directory


## Low level details
- Make generate sources the entire cluster/ directory from https://github.com/davidvossel/cluster-up
- Directly modifying cluster/ will cause the build to fail in travis. 
- cluster-hack/ directory is sourced from https://github.com/davidvossel/cluster-up and contains all the configs related to cluster-up and cluster-down
- cluster/examples moved to examples/
- Moved build logic from cluster/ephemeral-cluster-provider.sh to hack/build.sh
- A few new wrapper scripts have been introduced by the cluster-up project (cluster/ssh.sh, cluster/kubeconfig.sh, cluster/oc.sh)

The 'make cluster-up' and 'make cluster-down' logic has been split out into a separate repo located here, https://github.com/davidvossel/cluster-up


```release-note
Source "make cluster-up" logic from the new generic cluster-up repo. 
```

rmohr commented on hack/generate.sh:
```diff
@@ -47,3 +47,12 @@ protoc --go_out=plugins=grpc:. pkg/handler-launcher-com/cmd/info/info.proto
 mockgen -source pkg/handler-launcher-com/cmd/info/info.pb.go -package=info -destination=pkg/handler-launcher-com/cmd/info/generated_mock_info.go
+
+rm -rf cluster cluster-hack
+git clone https://github.com/davidvossel/cluster-up.git
```
@davidvossel can we just have this in the kubevirtci repo?

stu-gott replied:
This is the wrong approach. Please use a git subrepo instead of cloning.

davidvossel replied:
> This is the wrong approach. Please use a git subrepo instead of cloning.

git submodules?  I'd prefer we avoided submodules.   https://codingkilledthecat.wordpress.com/2012/04/28/why-your-company-shouldnt-use-git-submodules/

What would be the value of submodules over this initial approach. 

I don't think that there is anything inherently wrong with the approach as it is done here. It's not possible to modify the cluster/ dir directly without causing travis to fail. Just like we "generate" copy functions, schema's, etc... this is generating the cluster-up scripts.  If you update the hash env var, then you have to commit the results just like the rest of the generated code.

davidvossel replied:
> @davidvossel can we just have this in the kubevirtci repo?

That's something we should discuss. I'm not convinced where these scripts should live quite yet. We need to discuss how to run CI against this logic externally of kubevirt/kubevirt as well.

cynepco3hahue replied:
Imho it should be under the KubeVirt CI, in this case, update the images hashes under providers will be much easier(but in this case you will need to clone the whole kubevitrci repository)

rmohr replied:
We can do a shallow clone `git clone --depth 1`. Yes, let's discuss this. For instance in kubernetes the clonse in prow their test-infra repo on every test run.

davidvossel replied:
@cynepco3hahue @rmohr 

I've moved cluster-up logic to kubevirtci. https://github.com/kubevirt/kubevirtci/pull/87

davidvossel commented:
fixed travis issues. I hadn't fully separated which project owns which env vars in the config correctly when I split the orignial cluster/ dir into the cluster-up repo.

cynepco3hahue commented:
/retest

davidvossel commented:
we're waiting on https://github.com/kubevirt/kubevirtci/pull/87 to get merged into kubevirt ci before this PR will be finalized.

The logic around how we vendor the cluster-up code will be revised shortly

davidvossel commented on hack/generate.sh:
```diff
@@ -47,3 +47,7 @@ protoc --go_out=plugins=grpc:. pkg/handler-launcher-com/cmd/info/info.proto
+
+rm -rf cluster-up
+#TODO move this to official kubevirt/kubevirtci repo
+curl -L https://github.com/davidvossel/kubevirtci/archive/${kubevirtci_git_hash}/kubevirtci.tar.gz | tar xz kubevirtci-${kubevirtci_git_hash}/cluster-up --strip-component 1
```
@rmohr here's the logic we discussed. I've moved all the configs and dynamic data generated by cluster-up into a separate directory entirely (defaults to _ci-configs). THis means nothing is lost by deleting cluster-up/ and re-creating it during 'make generate'

for more info on the config's dir change, see this kubevirtci PR, https://github.com/kubevirt/kubevirtci/pull/88
